### PR TITLE
Get data upload status

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-713

#### Links to any Pull Requests related to this

UI - https://github.com/biomage-ltd/ui/pull/245
API - https://github.com/biomage-ltd/api/pull/105
IAC - https://github.com/biomage-ltd/iac/pull/110

#### Anything else the reviewers should know about the changes here

- Created new actions to save project and samples data to DynamoDB
- Staging in UI repo

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [X] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/specs/api.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ info:
   description: The main Node.js pipeline API.
   license:
     name: MIT
-    url: "https://github.com/biomage-ltd/api/blob/master/LICENSE"
```
Comment (xverges) [inline src/specs/api.yaml:8]: @aerlaut I thought that you and some other team member had conflicting setting to edit these yaml files, but I saw that it is you against you :-) https://github.com/biomage-ltd/api/commit/91eee5841b8967338e96d1ad21db240b3a9ec2a1#diff-9590aca4a81dce3ffd13c1d719b7c87a7e4c3e6a82b8fbd930e3ab1da22b4aba

Can you try to find out why this recurring `"` -> `'` -> `"` -> ... happens? Maybe we need to add some setting to the workspace?
https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml <- this is the extension that is recommended in our onboarding docs, and has a single quote setting

File: package.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
   "description": "Node.js API for the single-cell pipeline.",
   "main": "index.js",
   "scripts": {
-    "test": "jest --silent",
+    "test": "jest",
     "test:loud": "jest",
```
Comment (xverges) [inline package.json]: Do you think "loud" should be the default? If that is the case, should there be a `test:quiet` instead of the current `test:loud`?

File: src/api/route-services/__mocks__/projects.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  ) {
+    throw new Error('Invalid body');
+  }
+
+  return Promise.resolve({
+    name: 'Test project',
+    description: '',
+    createdDate: '',
+    lastModified: '',
+    uuid: 'project-1',
```
Comment (xverges) [inline src/api/route-services/__mocks__/projects.js]: Shouldn't this be the `projectUuid` that we got as a parameter?

File: src/specs/api.yaml [hunk 4]
```diff
@@ -405,16 +405,109 @@ paths:
...
+        '404':
+          description: Invalid projectUuid.
+      description: Create or update a project
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: ./models/api-body-schemas/Project.v1.yaml
+        description: The new configuration to update the old one by.
+  '/projects/{projectUuid}/samples':
```
Comment (xverges) [inline src/specs/api.yaml]: Aren't 404 status also needed for operations on sample resources?

File: src/specs/models/api-body-schemas/Project.v1.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+type: object
+title: ''
+description: Schema for a sample object
```
Comment (xverges) [inline src/specs/models/api-body-schemas/Project.v1.yaml]: Wrong description

File: tests/api/routes/samples.test.js [hunk 6]
```diff
@@ -24,18 +21,69 @@ describe('tests for samples route', () => {
...
+      .end((err) => {
+        if (err) {
+          return done(err);
+        }
+        return done();
+      });
+  });
+
+
+  it('Updating samples send error 500 if body is invalid', async (done) => {
```
Comment (xverges) [inline tests/api/routes/samples.test.js]: `500` indicates that the server is broken. When the request is not well constructed, we need to let the client know. A `400` would be more appropriate https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses 

In fact, this is being done for projects

**Review by xverges — CHANGES_REQUESTED** [review]
Do not return status 500 for invalid input

File: src/specs/api.yaml [hunk 7]
```diff
@@ -405,16 +405,109 @@ paths:
...
+                  type: string
+                samples:
+                  $ref: ./models/api-body-schemas/Samples.v1.yaml
+    get:
+      summary: 'Get samples under a project '
+      operationId: getProjectSamples
+      x-eov-operation-id: samples#get
+      x-eov-operation-handler: routes/samples
+      responses:
+        '200':
```
Comment (ivababukova) [inline src/specs/api.yaml:497]: the response can also be 404 if the sample is not found

File: src/specs/models/api-body-schemas/Project.v1.yaml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+properties:
+  name:
+    type: string
+  description:
+    type: string
+  createdDate:
+    type: string
+  lastModified:
+    type: string
+  uuid:
```
Comment (ivababukova) [inline src/specs/models/api-body-schemas/Project.v1.yaml:13]: I would call this `projectUuid` to make it less ambigious

File: tests/api/route-services/projects.test.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    const getItemSpy = jest.fn((x) => x);
+    AWSMock.setSDKInstance(AWS);
+    AWSMock.mock('DynamoDB', 'updateItem', (params, callback) => {
+      getItemSpy(params);
+      callback(null, dynamodbData);
+    });
+    return getItemSpy;
+  };
+
+  it('udpateProject work', async (done) => {
```
Comment (ivababukova) [inline tests/api/route-services/projects.test.js]: typos here

File: tests/api/route-services/samples.test.js [hunk 10]
```diff
@@ -54,15 +83,46 @@ describe('tests for the samples service', () => {
...
           TableName: 'samples-test',
-          Key: { experimentId: { S: '12345' } },
+          Key: { experimentId: { S: 'project-1' } },
           ProjectionExpression: 'samples.ids',
         });
       })
       .then(() => done());
   });
+
+  it('udpateSamples work', async (done) => {
```
Comment (ivababukova) [inline tests/api/route-services/samples.test.js]: typo in `udpateSamples ` too

Comment (ivababukova) [reply tests/api/route-services/samples.test.js]: apart from unit test for whether the update works, we should test for the behaviour for any corner cases that are possible, for example: What if we try to update non-existent sample? What if we pass empty samples entry in the update: is it going to delete all existent samples? What if dynamoDB was down and the actual update didn't happen?

File: tests/api/routes/projects.test.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    request(app)
+      .put('/v1/projects/someId')
+      .send(payload)
+      .expect(200)
+      .end((err) => {
+        if (err) {
+          console.log(err);
+          return done(err);
+        }
+        // there is no point testing for the values of the response body
```
Comment (ivababukova) [inline tests/api/routes/projects.test.js]: is it worth testing that if there is something wrong, the schema will catch it though? Like, write a test that passes misformatted input and expects the schema validation to fail

File: src/api/route-services/samples.js [hunk 12]
```diff
@@ -9,19 +9,18 @@ class SamplesService {
...
-    });
-
+  async getSamples(projectUuid) {
     const params = {
       TableName: this.tableName,
-      Key: key,
+      KeyConditionExpression: 'projectUuid = :projectUuid',
+      ExpressionAttributeValues: {
+        ':projectUuid': { S: projectUuid },
+      },
```
Comment (ivababukova) [inline src/api/route-services/samples.js]: It might make more sense to reuse the dynamodb marshall and unmarshall functions for this instead of manually formatting the input yourself

File: src/api/route-services/samples.js [hunk 13]
```diff
@@ -52,6 +51,31 @@ class SamplesService {
...
+    const marshalledData = convertToDynamoDbRecord({
+      ':samples': body.samples,
+      ':projectUuid': projectUuid,
+    });
+
+    // Update samples
+    const params = {
+      TableName: this.tableName,
+      Key: {
+        experimentId: { S: body.experimentId },
```
Comment (ivababukova) [inline src/api/route-services/samples.js]: It might make more sense to reuse the dynamodb marshall and unmarshall functions for this instead of manually formatting the input yourself

File: package.json (hunk 2 above)
Comment (aerlaut) [reply package.json]: Aah I see. I didn't see the `test:loud` before. I'll return this.

File: src/api/route-services/__mocks__/projects.js (hunk 3 above)
Comment (aerlaut) [reply src/api/route-services/__mocks__/projects.js]: No, the `projectUuid` inside a `project` is only `uuid`, so that you wouldn't be writing `project.projectUuid`.

File: tests/api/routes/samples.test.js (hunk 6 above)
Comment (aerlaut) [reply tests/api/routes/samples.test.js]: Ok, 400 seems a more fitting error. I'll change it

File: src/specs/models/api-body-schemas/Project.v1.yaml (hunk 8 above)
Comment (aerlaut) [reply src/specs/models/api-body-schemas/Project.v1.yaml:13]: As things currently are, `uuid` for an object (e.g. project or samples) refer to itself, so we wouldn't be doing `project.projectUuid` or `samples.sampleUuid`. This is how things are currently done in the API and UI. If we're going to make this change, then it's better change both in the UI and API. What do you think?

File: src/specs/api.yaml (hunk 7 above)
Comment (aerlaut) [reply src/specs/api.yaml:497]: Ok, will return this

File: tests/api/routes/projects.test.js (hunk 11 above)
Comment (aerlaut) [reply tests/api/routes/projects.test.js]: I think the tests for routes only checks for the existence and correct HTTP response of the routes, but does not check for what happens inside it, which is handed with tests in `route-services`.

File: src/api/route-services/samples.js (hunk 12 above)
Comment (aerlaut) [reply src/api/route-services/samples.js]: Yea, you're right, it's a risk to do it manually. I'll change it to use AWS's marshall.

File: src/specs/api.yaml (hunk 1 above)
Comment (aerlaut) [reply src/specs/api.yaml:8]: After checking`api.code-workspace`, there doesn't seem to be a rule enforcing single or double quote. I'm putting. We're using single quotes for JS files, so I'll put that rule in for YAML too.

File: src/specs/api.yaml (hunk 7 above)
Comment (aerlaut) [reply src/specs/api.yaml:497]: Added in new commit

File: src/specs/models/api-body-schemas/Project.v1.yaml (hunk 8 above)
Comment (ivababukova) [reply src/specs/models/api-body-schemas/Project.v1.yaml:13]: let's leave it as uuid then

File: src/api/route-services/projects.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+} = require('../../utils/dynamoDb');
+
+const { OK, NotFoundError } = require('../../utils/responses');
+
+class ProjectsService {
+  constructor() {
+    this.tableName = `projects-${config.clusterEnv}`;
+  }
+
+  async updateProject(projectUuid, project) {
```
Comment (ivababukova) [inline src/api/route-services/projects.js:40]: what is `project` equal to? Is it project name?

File: src/api/route-services/samples.js [hunk 15]
```diff
@@ -9,37 +11,39 @@ class SamplesService {
...
+      KeyConditionExpression: 'projectUuid = :projectUuid',
+      ExpressionAttributeValues: marshalledData,
       ProjectionExpression: 'samples',
     };
     const dynamodb = createDynamoDbInstance();
 
-    const response = await dynamodb.getItem(params).promise();
+    const response = await dynamodb.query(params).promise();
 
     if (response.Item) {
```
Comment (ivababukova) [inline src/api/route-services/samples.js:32]: Is there a chance that `response` will be null/none? Because if that happens, line 30 will crash and we should handle that case.

Comment (ivababukova) [reply src/api/route-services/samples.js:32]: the same situation on line 51 in this file, but I can't add a comment there because it hasn't been changed

File: src/api/route-services/samples.js [hunk 16]
```diff
@@ -9,37 +11,39 @@ class SamplesService {
...
       return prettyResponse;
     }
 
-    throw Error('Sample not found');
+    throw new NotFoundError('Samples not found');
   }
 
-  async getSampleIds(experimentId) {
-    const key = convertToDynamoDbRecord({
+  async getByExperimentId(experimentId) {
```
Comment (ivababukova) [inline src/api/route-services/samples.js]: if you are returning samples for an experiment, it is worth renaming that function to something like `getSampleIdsForExperiment` to make it more explicit

File: src/api/route-services/samples.js [hunk 17]
```diff
@@ -50,7 +54,36 @@ class SamplesService {
...
+    const marshalledKey = convertToDynamoDbRecord({
+      experimentId: body.experimentId,
+    });
+
+    const marshalledData = convertToDynamoDbRecord({
+      ':samples': body.samples,
+      ':projectUuid': projectUuid,
+    });
+
+    // Update samples
```
Comment (ivababukova) [inline src/api/route-services/samples.js:76]: do we have samples copied across different tables? Because if we do, then we will need to update all places where we store samples

File: src/api/route-services/projects.js (hunk 14 above)
Comment (ivababukova) [inline src/api/route-services/projects.js:40]: do we have project data copied across different tables? Because if we do, then we will need to update all places where we store projects, so that tables don't get out of sync

File: src/utils/responses/index.js [hunk 18]
```diff
@@ -0,0 +1,7 @@
+const NotFoundError = require('./NotFoundError');
```
Comment (ivababukova) [inline src/utils/responses/index.js:1]: what if there is a 5xx error?

File: tests/api/route-services/projects.test.js [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+      Attributes: AWS.DynamoDB.Converter.marshall({ projects: jsData }),
+    };
+    const getItemSpy = jest.fn((x) => x);
+    AWSMock.setSDKInstance(AWS);
+    AWSMock.mock('DynamoDB', 'updateItem', (params, callback) => {
+      getItemSpy(params);
+      callback(null, dynamodbData);
+    });
+    return getItemSpy;
+  };
```
Comment (ivababukova) [inline tests/api/route-services/projects.test.js]: we have this piece of code everywhere where we need to mock DynamoDB -- it might be worth moving it to a single place and reusing it in this and other test files.

Comment (ivababukova) [reply tests/api/route-services/projects.test.js]: you can check in jest documentation how you can set it up. Maybe this: https://jestjs.io/docs/mock-functions#mocking-modules and this: https://jestjs.io/docs/manual-mocks and this: https://stackoverflow.com/questions/49685265/jest-mock-user-module-in-all-test-files articles can be of help

File: src/api/route-services/projects.js (hunk 14 above)
Comment (aerlaut) [reply src/api/route-services/projects.js:40]: a `project` object that contains `uuid`, `samples`, etc.

Comment (aerlaut) [reply src/api/route-services/projects.js:40]: No, `project` data is only in `projects` table

File: src/utils/responses/index.js (hunk 18 above)
Comment (aerlaut) [reply src/utils/responses/index.js:1]: Currently it's not custom handled - the error just bubble up to the middleware. According to the error handling guide https://expressjs.com/en/guide/error-handling.html, errors which are 4xx and 5xx are bubbled accordingly, while errors outside these ranges are turned into error 500.

File: src/api/route-services/samples.js (hunk 17 above)
Comment (aerlaut) [reply src/api/route-services/samples.js:76]: No, `project` data is only in `projects` table

File: src/api/route-services/samples.js (hunk 16 above)
Comment (aerlaut) [reply src/api/route-services/samples.js]: As suggested, changed to `getSamplesByExperimentId`

File: src/utils/responses/index.js (hunk 18 above)
Comment (aerlaut) [reply src/utils/responses/index.js:1]: This `NotFoundError` is a custom error to handle specific conditions, so I'm not trying to catch all errors.

File: tests/api/route-services/projects.test.js (hunk 19 above)
Comment (aerlaut) [reply tests/api/route-services/projects.test.js]: Refactored into a separate module `test-utils/mockAWSServices.js`. Provides general test modules, custom behavior still has to be written manually.

File: src/api/route-services/samples.js (hunk 15 above)
Comment (aerlaut) [reply src/api/route-services/samples.js:32]: Hmm... I don't think so. It should return an error `503` if there's no connection. This should be thrown at await, so it shouldn't continue to line 30. The error would bubble up through the middleware.

File: tests/api/route-services/projects.test.js (hunk 19 above)
Comment (ivababukova) [reply tests/api/route-services/projects.test.js]: sounds great!

File: src/utils/responses/index.js (hunk 18 above)
Comment (ivababukova) [reply src/utils/responses/index.js:1]: ah cool, makes sense!

File: src/api/route-services/samples.js (hunk 15 above)
Comment (ivababukova) [reply src/api/route-services/samples.js:32]: in this case, we should be ok.

**Review by ivababukova — APPROVED** [review]
looks good :) It's merging has to be coordinated with the merging of the UI though: https://github.com/biomage-ltd/ui/pull/245/files

File: tests/test-utils/mockAWSServices.js [hunk 20]
```diff
@@ -0,0 +1,83 @@
+const AWSMock = require('aws-sdk-mock');
```
Comment (xverges) [inline tests/test-utils/mockAWSServices.js:1]: Love that you extracted this!

File: tests/test-utils/mockAWSServices.js [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+const AWS = require('../../src/utils/requireAWS');
+
+const mockDynamoGetItem = (payload = {}, error = null) => {
+  const dynamodbData = {
+    Item: AWS.DynamoDB.Converter.marshall(payload),
+  };
+  const getItemSpy = jest.fn((x) => x);
```
Comment (xverges) [inline tests/test-utils/mockAWSServices.js]: Minor issue: `getItemSpy` name has been copied to the rest of non-getItem methods. Maybe you can rename all to `spy` or similar?

**Review by xverges — APPROVED** [review]
See minor concern about variable names in mock methods

File: tests/test-utils/mockAWSServices.js (hunk 21 above)
Comment (aerlaut) [reply tests/test-utils/mockAWSServices.js]: changed to `fnSpy`
